pkg/csi/service/wcpguest/controller.go

```diff
@@ -452,6 +467,91 @@ func (c *controller) ControllerUnpublishVolume(ctx context.Context, req *csi.Con
+		if accessMode == v1.ReadWriteMany || accessMode == v1.ReadOnlyMany {
+			isFileVolume = true
+		}
+	}
+	// Check the feature state for file volume support
+	if isFileVolume {
+		if commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.FileVolume) {
+			// Retrieve Supervisor PVC
+
+			cfg, err := common.GetConfig(ctx)
+			if err != nil {
+				msg := fmt.Sprintf("failed to read config. Error: %+v", err)
+				log.Error(msg)
+				return nil, status.Errorf(codes.Internal, msg)
+			}
+			restClientConfig := k8s.GetRestClientConfigForSupervisor(ctx, cfg.GC.Endpoint, cfg.GC.Port)
+			// Adding watch on the CnsFileAccessConfig instance to register for updates
+			cnsFileAccessConfigWatcher, err := k8s.NewCnsFileAccessConfigWatcher(ctx, restClientConfig, c.supervisorNamespace)
+			if err != nil {
+				msg := fmt.Sprintf("failed to create cnsFileAccessConfigWatcher. Error: %+v", err)
+				log.Error(msg)
+				return nil, status.Errorf(codes.Internal, msg)
+			}
+			// Attacher timeout, default is set to 4 minutes
+			timeoutSeconds := int64(getAttacherTimeoutInMin(ctx) * 60)
+			cnsFileAccessConfigInstanceName := virtualMachine.Name + "-" + req.VolumeId
+			watchCnsFileAccessConfig, err := cnsFileAccessConfigWatcher.Watch(metav1.ListOptions{
+				FieldSelector:  fields.SelectorFromSet(fields.Set{"metadata.name": cnsFileAccessConfigInstanceName}).String(),
+				TimeoutSeconds: &timeoutSeconds,
+			})
+			if err != nil {
+				msg := fmt.Sprintf("failed to watch cnsFileAccessConfig instance %q with Error: %v", cnsFileAccessConfigInstanceName, err)
+				log.Error(msg)
+				return nil, status.Errorf(codes.Internal, msg)
+			}
+			if err := c.cnsOperatorClient.Delete(ctx, &cnsfileaccessconfigv1alpha1.CnsFileAccessConfig{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      cnsFileAccessConfigInstanceName,
+					Namespace: c.supervisorNamespace,
+				},
+			}); err != nil {
+				// Delete() on the CnsFileAccessConfig instance failed with different error
+				msg := fmt.Sprintf("failed to delete CnsFileAccessConfig instance: %q. Error: %+v", cnsFileAccessConfigInstanceName, err)
+				log.Error(msg)
+				return nil, status.Errorf(codes.Internal, msg)
+			}
+
+			defer watchCnsFileAccessConfig.Stop()
+			var cnsfileaccessconfigInstanceErr string
+			timeout := time.Now().Add(time.Duration(timeoutSeconds) * time.Second)
+			// Watch all update events made on CnsFileAccessConfig instance until Deleted event or a timeout occurs on the cnsfileaccessconfig instance
+			for {
+				log.Debugf("waiting for update on cnsfileaccessconfigs: %q", cnsFileAccessConfigInstanceName)
+				event := <-watchCnsFileAccessConfig.ResultChan()
+				cnsfileaccessconfig, ok := event.Object.(*cnsfileaccessconfigv1alpha1.CnsFileAccessConfig)
+				if !ok {
+					msg := fmt.Sprintf("Watch on cnsfileaccessconfig instance %q timed out", cnsFileAccessConfigInstanceName)
+					log.Error(msg)
+					return nil, status.Errorf(codes.Internal, msg)
+				}
+				if event.Type == "DELETED" {
```

Comment:
Shouldn;t you check the right cnsfileaccessconfig instance is deleted? From what I know, the watcher will give events for all cnsfileaccessconfig instances.

Reply:
Done